src/falcon/falcon_c/rpo.c

```diff
@@ -0,0 +1,757 @@
+    }
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        t1[i] = mult_mod_p(*(state + i), *(state + i));
+    }
+
+    uint64_t t2[STATE_WIDTH];
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        t2[i] = 0;
+    }
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        t2[i] = mult_mod_p(t1[i], t1[i]);
+    }
+
+    uint64_t t3[STATE_WIDTH];
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        t3[i] = 0;
+    }
+    exp_acc(3, t2, t2, t3);
+
+    uint64_t t4[STATE_WIDTH];
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        t4[i] = 0;
+    }
+    exp_acc(6, t3, t3, t4);
+
+    uint64_t tmp[STATE_WIDTH];
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        tmp[i] = 0;
+    }
+    exp_acc(12, t4, t4, tmp);
+
+    uint64_t t5[STATE_WIDTH];
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        t5[i] = 0;
+    }
+    exp_acc(6, tmp, t3, t5);
+
+    uint64_t t6[STATE_WIDTH];
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        t6[i] = 0;
+    }
+    exp_acc(31, t5, t5, t6);
+
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        uint64_t a = mult_mod_p(mult_mod_p(t6[i], t6[i]), t5[i]);
+        a = mult_mod_p(a, a);
+        a = mult_mod_p(a, a);
+        uint64_t b = mult_mod_p(mult_mod_p(t1[i], t2[i]), *(state + i));
+
+        *(state + i) = mult_mod_p(a, b);
+    }
+}
+
+void exp_acc(const uint64_t m, const uint64_t *base, const uint64_t *tail,
+             uint64_t *const res)
+{
+    for (uint64_t i = 0; i < m; i++)
+    {
+        for (uint64_t j = 0; j < STATE_WIDTH; j++)
+        {
+            if (i == 0)
+            {
+                *(res + j) = mult_mod_p(*(base + j), *(base + j));
+            }
+            else
+            {
+                *(res + j) = mult_mod_p(*(res + j), *(res + j));
+            }
+        }
+    }
+
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        *(res + i) = mult_mod_p(*(res + i), *(tail + i));
+    }
+}
+
+#include "rpo.h"
+
+/*
+ * Basic implementation of RPO128.
+ * Two TODOs remain:
+ * 1. Investigate the case of !FALCON_LE.
+ * 2. Investigate the padding rule in rpo128_inc_finalize as currently there are none.
+ */
+void rpo128_inc_init(inner_rpo128_context *rc)
+{
+    rc->dptr = 32;
+
+    /*
+     * Representation of an all-ones uint64_t is the same regardless
+     * of local endianness.
+     */
+    memset(rc->st.A, 0, sizeof rc->st.A);
+}
+
+void rpo128_inc_absorb(inner_rpo128_context *rc, const uint8_t *in, size_t len)
+{
+    size_t dptr;
+
+    dptr = (size_t)rc->dptr;
+    while (len > 0)
+    {
+        size_t clen, u;
+
+        /* 136 * 8 = 1088 bit for the rate portion in the case of SHAKE256
+            For RPO, this is 64 * 8 = 512 bits
+            The capacity for SHAKE256 is at the end while for RPO128 it is at the beginning */
+        clen = 96 - dptr;
+        if (clen > len)
+        {
+            clen = len;
+        }
+#if FALCON_LE // yyyLE+1
```

Comment:
What is `FALCON_LE`? I couldn't find it being mentioned anywhere else.

Reply:
This is related to the endianness. We will have to investigate its function in the original implementation but I have removed it for the time being.